server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java

```diff
@@ -444,6 +444,7 @@ private static void shutdown() {
             return; // never got far enough
         }
         var es = INSTANCE;
+        es.node.prepareForClose();
```

Comment:
3 lines further down there's a check for whether `es.node` is `null` before using it. Either that's not necessary - quite possible as I don't know the code well - or else this line needs to be guarded too.

Also, should this method be inside the `try` block? If it throws an exception then the `finally` block won't run, which presumably means Elasticsearch won't exit.

Reply:
The null check is outdated; it's not possible (I think it was at an earlier time). The INSTANCE null check is what guards early termination. However, moving this into the try though is important, so that any error while shutting down is properly logged and the latch counted down to allow the process to exit.